File: go/collatz-conjecture/collatz_conjecture.go
// Package collatzconjecture implements a solution for the collatz-conjecture exercism problem
package collatzconjecture

import (
	"fmt"
)

// CollatzConjecture returns the number of steps for the 3x+1 problem
func CollatzConjecture(n int) (int, error) {
	if n <= 0 {
		return -1, fmt.Errorf("defined only for n > 0, not %v", n)
	}
	steps := 0
	number := n
	for number > 1 {
		steps++
		if number%2 == 0 {
			number = number / 2
			continue
		}
		number = 3*number + 1
	}
	return steps, nil
}

File: python/collatz-conjecture/collatz_conjecture.py
def steps(number, number_of_steps=0):
    if number <= 0:
        raise ValueError(f"Defined only for positive numbers, not '{number}'.")

    if number == 1:
        return number_of_steps

    if number % 2 == 0:
        return steps(number / 2, number_of_steps + 1)

    return steps(3 * number + 1, number_of_steps + 1)


if __name__ == "__main__":
    print(f"{steps(120)=}")
    print(f"{steps(1000000)=}")
